File: test/extents_test.py
import pytest
from sys import float_info
import copclib as copc
from .utils import get_data_dir
import os


def test_copc_extents():

    point_format_id = 7
    num_eb_items = 3

    #### Constructor ####

    # Empty Constructor

    extents = copc.CopcExtents(point_format_id, num_eb_items)
    assert extents.point_format_id == point_format_id
    assert len(extents.extra_bytes) == num_eb_items

    # Point format checks
    with pytest.raises(RuntimeError):
        copc.CopcExtents(5)
        copc.CopcExtents(9)

    #### Default Extents ####

    file_path = os.path.join(get_data_dir(), "writer_test.copc.laz")

    eb_vlr = copc.EbVlr(num_eb_items)
    cfg = copc.CopcConfigWriter(point_format_id, extra_bytes_vlr=eb_vlr)
    writer = copc.FileWriter(file_path, cfg)

    extents = writer.copc_config.copc_extents
    assert extents.point_format_id == 7
    assert len(extents.extra_bytes) == 3

    writer.Close()

    reader = copc.FileReader(file_path)

    extents = reader.copc_config.copc_extents

    assert extents.point_format_id == 7
    assert extents.has_extended_stats == False
    assert len(extents.extra_bytes) == 3

    assert extents.classification.minimum == 0
    assert extents.intensity.minimum == 0

    for extent in extents.extents:
        assert extent.minimum == 0
        assert extent.maximum == 0

    #### Set Extents ####
    cfg = copc.CopcConfigWriter(
        point_format_id, extra_bytes_vlr=eb_vlr, has_extended_stats=True
    )
    writer = copc.FileWriter(file_path, cfg)

    extents = writer.copc_config.copc_extents

    assert extents.has_extended_stats == True

    extents.x.minimum = 1
    extents.x.maximum = 2
    extents.x.mean = 3
    extents.x.var = 4
    extents.y.minimum = 5
    extents.y.maximum = 6
    extents.y.mean = 7
    extents.y.var = 8
    extents.z.minimum = 9
    extents.z.maximum = 10
    extents.z.mean = 11
    extents.z.var = 12

    extents.classification.minimum = -float_info.max
    extents.classification.maximum = float_info.max
    extents.classification.mean = 15
    extents.classification.var = 5

    extents.intensity = (-10, 5)  # Tuple implicit conversion

    extents.extra_bytes[0].minimum = -float_info.max
    extents.extra_bytes[0].maximum = float_info.max
    extents.extra_bytes[0].mean = 566
    extents.extra_bytes[0].var = 158

    writer.Close()

    reader = copc.FileReader(file_path)

    extents = reader.copc_config.copc_extents

    assert extents.x.minimum == 1
    assert extents.x.maximum == 2
    assert extents.x.mean == 3
    assert extents.x.var == 4
    assert extents.y.minimum == 5
    assert extents.y.maximum == 6
    assert extents.y.mean == 7
    assert extents.y.var == 8
    assert extents.z.minimum == 9
    assert extents.z.maximum == 10
    assert extents.z.mean == 11
    assert extents.z.var == 12

    assert extents.classification.minimum == -float_info.max
    assert extents.classification.maximum == float_info.max
    assert extents.classification.mean == 15
    assert extents.classification.var == 5

    assert extents.intensity == (-10, 5, 0, 1)  # Tuple implicit conversion
    assert extents.intensity == (-10, 5)  # Tuple implicit conversion

    assert extents.extra_bytes[0].minimum == -float_info.max
    assert extents.extra_bytes[0].maximum == float_info.max
    assert extents.extra_bytes[0].mean == 566
    assert extents.extra_bytes[0].var == 158

    #### Extra byte list setter ####
    writer = copc.FileWriter(file_path, cfg)
    extents = writer.copc_config.copc_extents

    with pytest.raises(RuntimeError):
        extents.extra_bytes = [(0, 1), (2, 3)]
        extents.extra_bytes = [(0, 1), (2, 3), (4, 5), (6, 7)]

    extents.extra_bytes = [(0, 1), (2, 3), (4, 5)]

    writer.Close()

    reader = copc.FileReader(file_path)

    extents = reader.copc_config.copc_extents

    assert extents.extra_bytes[0].minimum == 0
    assert extents.extra_bytes[0].maximum == 1

    assert extents.extra_bytes[1].minimum == 2
    assert extents.extra_bytes[1].maximum == 3

    assert extents.extra_bytes[2].minimum == 4
    assert extents.extra_bytes[2].maximum == 5

    #### Get/Set Extents List ####
    extents = copc.CopcExtents(point_format_id, num_eb_items)

    for extent in extents.extents:
        extent.minimum = 1
        extent.maximum = 1

    for extent in extents.extents:
        assert extent.minimum == 1
        assert extent.maximum == 1

File: test/extents_test.cpp
#include <limits>
#include <string>

#include <catch2/catch.hpp>
#include <copc-lib/copc/extents.hpp>
#include <copc-lib/io/copc_reader.hpp>
#include <copc-lib/io/copc_writer.hpp>
#include <copc-lib/las/vlr.hpp>

using namespace copc;
using namespace std;

TEST_CASE("COPC Extent", "[CopcExtent]")
{
    CopcExtent extent(-numeric_limits<double>::max(), numeric_limits<double>::max(), 5, 10);

    REQUIRE(extent.minimum == -numeric_limits<double>::max());
    REQUIRE(extent.maximum == numeric_limits<double>::max());
    REQUIRE(extent.mean == 5);
    REQUIRE(extent.var == 10);

    CopcExtent other(-numeric_limits<double>::max(), numeric_limits<double>::max(), 5, 10);
    REQUIRE(extent == other);

    other = CopcExtent(-numeric_limits<double>::max(), 5);

    REQUIRE(extent != other);

    REQUIRE_THROWS(CopcExtent(1, 0));

    SECTION("ToString")
    {
        extent.ToString();
        std::stringstream ss;
        ss << extent;
    }
}

TEST_CASE("COPC Extents", "[CopcExtents]")
{

    int8_t point_format_id{7};
    uint16_t num_eb_items{5};

    SECTION("Constructor")
    {
        // Empty constructor
        {
            CopcExtents extents{point_format_id, num_eb_items};
            REQUIRE(extents.PointFormatId() == point_format_id);
            REQUIRE(extents.ExtraBytes().size() == num_eb_items);
        }
        // Vlr Constructor
        {
            auto vlr = las::CopcExtentsVlr();
            vlr.items.resize(CopcExtents::NumberOfExtents(point_format_id, num_eb_items), {0, 0});
            CopcExtents extents{vlr, point_format_id, num_eb_items};
            REQUIRE(extents.PointFormatId() == point_format_id);
            REQUIRE(extents.ExtraBytes().size() == num_eb_items);
            REQUIRE(extents.Intensity()->minimum == 0);
            REQUIRE(extents.Intensity()->maximum == 0);
            REQUIRE(*extents.Intensity() == CopcExtent(0, 0)); // Test == operator
            REQUIRE(*extents.Intensity() != CopcExtent(0, 1)); // Test != operator
        }
        // Point format checks
        REQUIRE_THROWS(CopcExtents(5));
        REQUIRE_THROWS(CopcExtents(9));
    }

    SECTION("Default Extents")
    {
        string file_path = "writer_test.copc.laz";
        {
            las::EbVlr eb_vlr(2);
            eb_vlr.items[0].data_type = 29; // byte size 12
            eb_vlr.items[1].data_type = 29; // byte size 12

            CopcConfigWriter cfg(7, {}, {}, {}, eb_vlr);
            FileWriter writer(file_path, cfg);

            auto extents = writer.CopcConfig()->CopcExtents();
            REQUIRE(extents->PointFormatId() == 7);
            REQUIRE(extents->HasExtendedStats() == false);
            REQUIRE(extents->ExtraBytes().size() == 2);

            writer.Close();
        }
        {
            FileReader reader(file_path);

            auto extents = reader.CopcConfig().CopcExtents();

            REQUIRE(extents.PointFormatId() == 7);
            REQUIRE(extents.HasExtendedStats() == false);
            REQUIRE(extents.ExtraBytes().size() == 2);

            REQUIRE(extents.Intensity()->minimum == 0);
            REQUIRE(extents.Classification()->minimum == 0);

            for (const auto &extent : extents.Extents())
            {
                REQUIRE(extent->minimum == 0);
                REQUIRE(extent->maximum == 0);
            }
        }
    }

    SECTION("Set Extents")
    {
        string file_path = "writer_test.copc.laz";
        {
            las::EbVlr eb_vlr(1);
            eb_vlr.items[0].data_type = 29; // byte size 12

            CopcConfigWriter cfg(7, {}, {}, {}, eb_vlr, true);
            FileWriter writer(file_path, cfg);

            auto extents = writer.CopcConfig()->CopcExtents();

            REQUIRE(extents->HasExtendedStats() == true);

            extents->X()->minimum = -20;
            extents->X()->maximum = -10;
            extents->X()->mean = 20;
            extents->X()->var = 10;
            extents->Y()->minimum = -30;
            extents->Y()->maximum = -20;
            extents->Y()->mean = 10;
            extents->Y()->var = 50;
            extents->Z()->minimum = -10;
            extents->Z()->maximum = -6;
            extents->Z()->mean = 4;
            extents->Z()->var = 12;

            extents->Intensity()->minimum = -numeric_limits<double>::max();
            extents->Intensity()->maximum = numeric_limits<double>::max();
            extents->Intensity()->mean = 15;
            extents->Intensity()->var = 5;

            extents->Classification()->minimum = 0;
            extents->Classification()->maximum = 255;

            extents->ExtraBytes()[0]->minimum = -numeric_limits<double>::max();
            extents->ExtraBytes()[0]->maximum = numeric_limits<double>::max();
            extents->ExtraBytes()[0]->mean = 566;
            extents->ExtraBytes()[0]->var = 158;

            // Vector accessor
            extents->Extents()[10]->minimum = 78;
            extents->Extents()[10]->maximum = 79;

            extents->Extents()[11]->minimum = 80;
            extents->Extents()[11]->maximum = 81;

            writer.Close();
        }
        {
            FileReader reader(file_path);

            auto extents = reader.CopcConfig().CopcExtents();

            REQUIRE(extents.X()->minimum == -20);
            REQUIRE(extents.X()->maximum == -10);
            REQUIRE(extents.X()->mean == 20);
            REQUIRE(extents.X()->var == 10);
            REQUIRE(extents.Y()->minimum == -30);
            REQUIRE(extents.Y()->maximum == -20);
            REQUIRE(extents.Y()->mean == 10);
            REQUIRE(extents.Y()->var == 50);
            REQUIRE(extents.Z()->minimum == -10);
            REQUIRE(extents.Z()->maximum == -6);
            REQUIRE(extents.Z()->mean == 4);
            REQUIRE(extents.Z()->var == 12);

            REQUIRE(extents.Intensity()->minimum == -numeric_limits<double>::max());
            REQUIRE(extents.Intensity()->maximum == numeric_limits<double>::max());
            REQUIRE(extents.Intensity()->mean == 15);
            REQUIRE(extents.Intensity()->var == 5);

            REQUIRE(*extents.Classification() == CopcExtent(0, 255));
            REQUIRE(*extents.Classification() == CopcExtent(0, 255, 0, 1));

            REQUIRE(extents.ExtraBytes()[0]->minimum == -numeric_limits<double>::max());
            REQUIRE(extents.ExtraBytes()[0]->maximum == numeric_limits<double>::max());
            REQUIRE(extents.ExtraBytes()[0]->mean == 566);
            REQUIRE(extents.ExtraBytes()[0]->var == 158);

            // Vector accessor
            REQUIRE(extents.Extents()[10]->minimum == 78);
            REQUIRE(extents.Extents()[10]->maximum == 79);
            REQUIRE(*extents.UserData() == CopcExtent(78, 79));

            REQUIRE(extents.Extents()[11]->minimum == 80);
            REQUIRE(extents.Extents()[11]->maximum == 81);
            REQUIRE(*extents.ScanAngle() == CopcExtent(80, 81));
        }
    }

    SECTION("ToCopcExtentVlr")
    {

        CopcExtents extents{point_format_id, num_eb_items};

        auto vlr = extents.ToLazPerf({}, {}, {});
        REQUIRE(vlr.items.size() == CopcExtents::NumberOfExtents(point_format_id, num_eb_items));
    }

    SECTION("Get/Set Extents")
    {
        CopcExtents extents{point_format_id, num_eb_items};

        for (const auto &extent : extents.Extents())
        {
            extent->minimum = 1;
            extent->maximum = 1;
        }

        for (const auto &extent : extents.Extents())
        {
            REQUIRE(extent->minimum == 1);
            REQUIRE(extent->maximum == 1);
        }
    }
    // TODO[Leo]: Add all Extents functions to tests
}
